homeassistant/components/deconz.py

```diff
@@ -0,0 +1,134 @@
+"""Platform integrating Deconz support.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/deconz/
+"""
+
+import asyncio
+import logging
+import voluptuous as vol
+
+from homeassistant.const import (
+    CONF_API_KEY, CONF_HOST, CONF_PASSWORD, CONF_PORT, CONF_USERNAME,
+    EVENT_HOMEASSISTANT_STOP)
+from homeassistant.core import callback, EventOrigin
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers import discovery
+from homeassistant.util import slugify
+from homeassistant.util.json import load_json, save_json
+
+REQUIREMENTS = ['pydeconz==10']
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = 'deconz'
+
+DECONZ_DATA = 'deconz_data'
+CONFIG_FILE = 'deconz.conf'
+
+TYPE_AS_EVENT = 'type_as_event'
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_HOST): cv.string,
+        vol.Optional(CONF_API_KEY): cv.string,
+        vol.Optional(CONF_PORT, default=80): cv.port,
+        vol.Optional(TYPE_AS_EVENT, default=['ZHASwitch']):
+            vol.All(cv.ensure_list, [cv.string]),
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+
+@asyncio.coroutine
+def async_setup(hass, config):
+    """Setup services for Deconz."""
+    deconz_config = config[DOMAIN]
+    config_file = load_json(hass.config.path(CONFIG_FILE))
+
+    @callback
+    def _shutdown(call):  # pylint: disable=unused-argument
+        """Stop the connections to Deconz on shutdown."""
+        if DECONZ_DATA in hass.data:
+            _LOGGER.info("Stopping Deconz session.")
+            hass.data[DECONZ_DATA].close()
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, _shutdown)
+
+    @asyncio.coroutine
+    def generate_api_key(call):
+        """Generate API key needed to communicate with Deconz.
+
+        Store API key in deconz.conf.
+        """
+        from pydeconz.utils import get_api_key
+        deconz_config[CONF_USERNAME] = call.data.get(CONF_USERNAME, 'delight')
```

Comment:
I don't think that the default values for username / password should be hard coded here, but rather added as default value to the config schema. Nobody's going to look for the default value here. ;)

Reply:
I guess I could create a second config schema for the service? Since the username and password is only relevant for the service to generate the API key. Not sure what is best :)

Reply:
Oh, I just now realize that this is a service! Why do you make this a service? The component cannot work before that service was called / the API key was generated, can it? So how about this workflow:
* On startup, check if an API key was generated
  * If so, initialize the component
  * If not, retrieve username / password from the config, generate a new API key
    * If that worked, initialize the component
    * If it didn't, fail the initialization

Reply:
I had it like that initially, but then I thought that I don't want the user to need to remove the username and password afterwords. Originally I created a notification in the GUI for the user to write down and add to the configuration.yaml. This is the opposite in that after initially specifying the component you are not required to add any additional configuration for the component.

But in the end I guess it is a matter of taste.

Reply:
I'm gonna think a bit more about this, you have some valid points!

Reply:
I changed it to more as you suggested